testdata/p4_16_samples_outputs/psa-register-complex-bmv2-midend.p4

```diff
@@ -37,7 +37,7 @@ control cIngress(inout headers_t hdr, inout metadata_t user_meta, in psa_ingress
         regfile_0.write(32w1, 48w3);
         regfile_0.write(32w2, 48w4);
         tmp = regfile_0.read(32w1);
-        hdr.ethernet.dstAddr = tmp + regfile_0.read(32w2) + 48w281474976710651;
+        hdr.ethernet.dstAddr = regfile_0.read(32w1) + regfile_0.read(32w2) + 48w281474976710651;
```

Comment:
this is a weird one.
many passes assume that there aren't two operations with side-effects in the same statement.
so this one I would think should not happen

Reply:
The weird thing is that the mechanism used in the LocalCopyPropagation pass (the hasSideEffects class) to detect when a value of a variable has side-effects does not detect that the value of '_tmp_' variable here has any side-effects, this is the reason why it's value does get propagated into the following statement. Since the mechanism used here is the same for all of the passes (I assume the hasSideEffects class is responsible for this throughout the compiler) I don't think there would be any problems with other passes, unless I am missunderstanding something.

The reason this value is not propagated when the pass is run only once is because of the ordering of the statements in the program. Since the ordering be changed arbitrarily, the number of repeated runs of LocalCopyProp pass needed for this kind of behaviour to show can also vary. This is the reason why this situation did not appear before, or at least not consistently enough to be noticed.

If there is still a need to avoid these scenarios, an Inspector can be added in LocalCopyProp which is run on the right side of every AssignmentStatement and sets a flag if it detects that in that right side an ExternMethod exists, and then the LocalCopyPropagation pass can use this flag to determine if it should propagate further variables whose values are also a ExternMethod.

Reply:
SideEffects is very simple. We have to figure out why the expression reading the register file is not detected as having side effects. Do you have time to look at this? 

Mihai

Reply:
It seems to have a noSideEffects annotation attached to it.

Reply:
So the thing to do is to understand whether the meaning of this annotation is the same as what the compiler worries about. Let me think about it a little bit.

Reply:
This is from the spec:
```
@noSideEffects - Weaker than @pure and describes a function that does not change any hidden state, but may depend on hidden state. One example is a hash function that computes a deterministic hash of its arguments, plus some internal state that can be modified via control plane API calls such as a seed or initialization vector. Another example is a read of one element of a register array extern object. Such a function may be dead code eliminated, and may be reordered or combined with other @noSideEffects or @pure calls (subject to the limits imposed by copy-out behavior of out and inout parameters), but not with other function calls that may have side effects that affect the function. 
```

Reply:
If I understand correctly, we need to know if other function calls within a statement affect the function call we are about to propagate. This doesn't seem to be checked anywhere in the LocalCopyProp pass, so how would we achieve this?

Reply:
That is a vague statement unfortunately.
The problem is not with this pass, the work done by this pass is correct.
The problem is with other passes, that may expect that there aren't two method calls in the same statement.
It's not easy to review all other passes to make sure that this is what they expect.

Reply:
We can leave it like this and see whether anything breaks.
I will approve this for now.

Reply:
I was wondering, can BuiltInMethods be problematic as well when having multiple in one statement?

Reply:
BTW: this invariant is established by SideEffectOrdering, early in the compiler.
The invariant is that each statement has at most one side effect. 
But the definition of "side-effect" is subtle, as can be seen from the spec, where `@pure` and `@noSideEffect` are two forms of possible non-side effects. So it is not exactly clear what each pass may expect.

Reply:
What is different about BuiltInMethods is that the compiler fully understands what they do. But for extern methods the compiler only has a promise provided by their API and these annotations.

Reply:
Understood, would you like for me to push the changes for the init_apply function and then merge this or should we skip that as well?

Reply:
That is to some degree a cosmetic change. They are equivalent if you always start the visitor from the P4Program node. But if you invoke the visitor on a subtree, they are not equivalent. If you plan to do it I will wait.

Reply:
I will update the PR with that change and then we can move forward with the merge.